doc/hare-dtm-hld.md

```diff
@@ -0,0 +1,614 @@
+receives a failure event, it creates an event and enqueues it into the event
+queue (EQ). RC leader then processes the event. It updates the failure count
+with respect to that epoch and creates a new failure vector entry as mentioned
+above. While updating the failure vector, it is possible that there are duplicate
+event queue entries for the same event (as the failure event is received on
+multiple nodes). In order to handle this duplication, RC leader checks if the
+corresponding failure event is already added to the failure vector for the
+given epoch and discards the duplicate events if already recorded.
+Following sequence diagram illustrates the flow of information in this case,
+
+<!--
+@startuml
+autonumber
+ha -> hax0_RC: reports failure of Motr1
+ha -> hax1: reports failure of Motr1
+ha -> hax2: reports failure of Motr1
+hax0_RC -> Consul: Add a failure vector update event to EQ with epoch e
+hax1 -> Consul: Add a failure vector update event to EQ with epoch e
+hax2 -> Consul: Add a failure vector update event to EQ with epoch e
+hax0_RC -> Consul: process EQ
+hax0_RC -> hax0_RC: gets the event, updates the failure vector dict for the given epoch
+hax0_RC -> Consul: Fetch next event
+hax0_RC -> hax0_RC: discard the event if already recorded for the given epoch
+@enduml
+-->
+![img](failvec-update.png)
+
+Hax can then iterate over required epochs for a process to create a list of
+relevant failure events from the corresponding failure vectors.
+
+#### Failvec build
+During DTM recovery Hare needs to send all the ha notifications corresponding
+to a Motr process. This list of ha notifications can be built using the failure
+events saved in Consul with respect to a single or multiple epochs. Hax fetches
+KV(s) corresponding to current or a set of epochs. By iterating over the ha
+events dictionaries from each key, hax can build a single list of ha events
+to be sent to one or more Motr processes.
+
+
+### DTM Messages
+Hare sends the failure vector and ha states corresponding to a recovering
+process using Motr [ha_note](https://github.com/Seagate/cortx-Motr/blob/83f069b1b1c4954c3835d367c7bb214c51df4770/ha/note.h#L159)
+and halink protocol. All outgoing messages from Hare are sent over Motr [halink](https://github.com/Seagate/cortx-Motr/blob/main/ha/link.h)
+or via explicit Motr [Spiel](https://github.com/Seagate/cortx-Motr/blob/main/spiel/spiel.h)
+interface. Hare implements data structures that map to Motr DTM0 data structures.
+
+
+### DTM participants
+All the peer Motr processes from the pool corresponding to the failed process
+participate in the DTM recovery of that process. Thus Hare broadcasts all the
+relevant ha events and DTM messages to the participating processes only as
+depicted by the following sequence diagram.
+
+<!--
+@startuml
+autonumber
+m0client -> Motr1: IO
+m0client -> Motr2: IO
+m0client -> Motr3: IO
+HA -> Hare: Report Motr2 Node failure
+destroy Motr2
+Consul -> Hax: Report service failure
```

Comment:
Yes, so we either need to ignore Consul events or HA events or handle both and understand that the respective event is already applied or not. Because one can come after another and we need to understand which to ignore.
This we need to handle separately anyway and is not specifically related to DTM. We need another design for HA - Hare integration/interaction. 

>It seems like Hax should read as Hare in this diagram.
And that will highlight that we'll have two sources of truth: HA events and Consul Health updates.
At the moment, Notify FAILED are sent out as a result of Consul notification. But in this picture we have also HA.

I think to avoid confusion here, I will just mention one source of failure events, I thought of covering the HA aspect as well but that is out of scope of this document so better let's avoid that here. For now we just need someone to report the failure.